### ISSUE TITLE
Typescript Typings

### ISSUE THREAD

**Issue by rtablada** [san650/ember-cli-page-object#426]
I would like to see if there are any interests or existing efforts to add typings for Ember CLI Page Object to help work with Typescript support and general editor support and suggestions/autocompletes.

**Comment by leondmello** [comment]
Yea, this would be awesome 😄

**Comment by ro0gr** [comment]
Glad to see the ticket.

I've made [some](https://github.com/ro0gr/ember-cli-page-object/blob/9a5a6be96fc03f69b95edf3f3515aab8bfab42f7/addon-test-support/ts/po.ts) progress with definitions recently.

It auto-completes sub-components and attrs:

```ts
const p = create({
  scope: '.AComponent',

  b: { ... },

  doSomething: clickable()
});

a.b.isVisible;
a.b.click();
a.doSomething().b.focus();

// it fails:
a.isVisible.b
a.isVisible.click();
a.click().isVisible
``` 

There is still lots of work to be done yet. 

TS with pojos is problematic. Custom PO methods have `this: any` by default. Maybe public typing for component should be providedm like

```
export { Component } from 'ember-cli-page-object';

const Def = {
  doIt(this: Component) {
  }
};
```

Will try to come up with some PR.

**Comment by san650** [comment]
I'm not a typescript user (yet) but this looks amazing 👏👏

**Comment by NullVoxPopuli** [comment]
This is incomplete, but this is what I have so far:
```ts
// types/ember-cli-page-object/index.d.ts
type Options = {
  multiple?: boolean;
};

export function create(page: object): any;
export function collection(selector: string, page: object): any;
export function clickable(selector?: string): any;
export function isVisible(selector?: string, options?: Options): boolean;
export function fillable(selector?: string): any;
export function text(selector?: string): string;
export function count(selector?: string): number;
export function is(selector?: string): boolean;
export function property(selectorOrProperty?: string, selector?: string): string;
export function hasClass(className: string): boolean;
export function isPresent(selector?: string): boolean;

// types/ember-cli-page-object/macros.d.ts
export function getter<T>(fn: () => T): T;
```

**Comment by ro0gr** [comment]
@NullVoxPopuli you might be interested in https://github.com/san650/ember-cli-page-object/pull/458

**Comment by ro0gr** [comment]
Ambient types was included to 1.16.0 release, thank you all for the patience!
